lib/skate/detours/missed_stops.ex

```diff
@@ -0,0 +1,128 @@
+alias Util.Location
+
+defmodule Skate.Detours.MissedStops do
+  @moduledoc """
+  Processes Shape and Stop data to compute contiguous stops between the
+  `connection_start` and `connection_end` parameters.
+  """
+
+  @type t :: %__MODULE__{
+          connection_start: Util.Location.From.t(),
+          connection_end: Util.Location.From.t(),
+          stops: nonempty_list(Util.Location.From.t()),
+          shape: nonempty_list(Util.Location.From.t())
+        }
+
+  @enforce_keys [:connection_start, :connection_end, :stops, :shape]
+  defstruct [:connection_start, :connection_end, :stops, :shape]
+
+  @doc """
+  Returns the contiguous list of stops, from the input parameter `cfg.stops`.
+  """
+  @spec missed_stops(cfg :: Skate.Detours.MissedStops.t()) :: [Util.Location.From.t()]
+  def missed_stops(%__MODULE__{
+        stops: stops,
+        shape: shape,
+        connection_start: connection_start,
+        connection_end: connection_end
+      }) do
+    segmented_shape = segment_shape_by_stops(shape, stops)
+
+    {connection_start, connection_end}
+    |> missed_segments(segmented_shape)
+    |> Enum.flat_map(& &1.stops)
+  end
+
+  defp missed_segments({connection_start, connection_end}, segmented_shape) do
+    %{index: start_index} = get_index_by_min_dist(segmented_shape, connection_start)
+    %{index: end_index} = get_index_by_min_dist(segmented_shape, connection_end)
+
+    if start_index == end_index do
+      %Skate.Detours.ShapeSegment{
+        # As long as there's only one stop in the segment,
+        # we can assume that we won't miss any stops
+        # because we're rejoining the same segment
+        stops: [_]
+      } = Enum.at(segmented_shape, start_index)
+
+      []
+    else
+      # We don't want the stop associated with the segment that the detour rejoins.
+      # So subtract one from the end index to ignore that segment
+      Enum.slice(segmented_shape, start_index..(end_index - 1))
+    end
+  end
+
+  defp segment_shape_by_stops([] = _shape, [] = _stops),
```

Comment:
**suggestion:** it would probably be more for documentation than for actual type-safety, but I think a typespec for this function might help make it slightly more readable.

Reply:
I think the same probably goes for a couple of the other private functions in this module.

Reply:
PTALA 572fa5784ae87befbfc81532f905843a93e7ca25 🙏🏻

Reply:
Awesome, thanks!